engine/equivalence_source_target.ml

```diff
@@ -35,13 +35,26 @@ let constructor_to_domain repr_env : constructor -> domain = function
      | Tag t -> Domain.tag (Domain.Set.point t)
 
 let constrained_subtrees repr_env children fallback =
+  let head_domain kst_list =
+    let _head_domain : constructor -> Domain.t = function
+      | String _ -> failwith "Not implemented"
+      (* | Variant _v -> assert false *)
+      | Nil -> Domain.empty
```

Comment:
Be careful that `Nil` is not a `nil` value like in Python but the empty list, so you should have a case for `| Nil | Cons _` that returns the union of `int 0` (for `Nil`) and `tag 0` (for `Cons`).